rfcs/005 - Unsurprising IaSQL Functions RFC.md

```diff
@@ -0,0 +1,174 @@
+- Consumes *far* more data in the database as the SQL statements are kept for potentially months until eventually dropped from the job table.
+- More likely to need a split of complex RPC calls into multiple sub-calls, as no good way to establish a true socket-like communication channel.
+
+##### Extension attaching a second PSQL socket to process
+
+Instead of recreating the RPC mechanism or emulating it, the IaSQL functions call an extension that does two things:
+
+1. Spin up a random port exposing a second PSQL protocol socket.
+2. Calling a REST endpoint on the engine passing the function called and arguments, along with the new socket number.
+
+The extension then polls for:
+
+1. A connection being established on that socket.
+2. Executing whatever SQL comes in on that socket (perhaps as simple as passing this off to Postgres' own internals).
+3. The connection closing.
+
+Then it disables the socket and simply exits the extension function (the wrapper function that called it does what it wants after that to potentially return metadata results to the caller).
+
+**Pros:**
+
+- The `scheduler` process is no longer necessary, but we still get to re-use the vast majority of our current code.
+- No weird injection in TypeORM to capture it's output, just configuring it to talk to a different port than normal.
+
+**Cons:**
+
+- Need to write a Rust extension, and faking the standard postgres connection may be more difficult than expected (it may need to interpret and inject into the main connection)
+- Need to expose all ports between the Postgres server and the IaSQL engine, and RDS cannot be used
+
+#### Change Postgres Behavior
+
+##### Fork Postgres, make it like MySQL
+
+If we forked postgres itself, we could eliminate the transaction-like behavior of connections and this problem goes away.
+
+**Pros:**
+
+- Solves the issue, [but at what cost?](https://i.kym-cdn.com/entries/icons/mobile/000/032/562/iplcoialdcw31.jpg)
+
+**Cons:**
+
+- Very hard to do in a codebase we are unfamiliar with
+- Negatively impacts expected Postgres behavior in all other circumstances
+
+##### Use `pgbouncer` with `statement`-level `pool_mode`
+
+`pgbouncer` is a commonly-used service in conjunction with Postgres to allow thousands of worker processes connect to Postgres without exploding the server with too many child processes. [The `statement`-level in its `pool_mode` configuration](https://www.pgbouncer.org/config.html#pool_mode) bans requests that span more than one statement (and is likely why ArcType and other database clients simply split the requests by statement and issue them separately).
+
+**Pros:**
+
+- "Solves" the issue, by breaking connections that try to do multi-statement at all.
+
+**Cons:**
+
+- "Solves" the issue, by breaking connections that try to do multi-statement at all.
+
+##### Implement a Postgres proxy in the Engine
```

Comment:
that sounds as a reasonable approach to me... we should focus on minimizing the load + latency

Reply:
I am also leaning towards it being the first thing we do, also because it lets us sidestep an issue with the current RDS approach (no SSL cert), but I have realized after writing that up that there's still going to be an issue if the user tries to write their query in a real transaction -- that would be sent to the database as a singular "statement" but would run into the same data issues as before.